{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3be1d41e-2365-4c09-84c2-84e177cf254c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Metagenomics Bioinformatics Course - EBI MGnify 2021\n",
    "## MGnify Services and API - Practical exercise\n",
    "\n",
    "### Aims\n",
    "In this exercise, we will learn how to use the [Mgnify API](https://www.ebi.ac.uk/metagenomics/api/v1).\n",
    "\n",
    "- Discover the available data on the MGnify website\n",
    "- Learn how to use the MGnify API to fetch data using scripts or analysis notebooks\n",
    "\n",
    "### How this works\n",
    "This file is a [Jupyter Notebook](https://jupyter.org). \n",
    "It has instructions, and also code cells. The code cells are connected to Python, and you can run all of the code in a cell by pressing Play (▶) icon in the top bar, or pressing `shift + return`.\n",
    "The code libraries you should need are already installed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9704e24e-b25c-4c5e-b02d-ecd4f23d085c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[pandas](https://pandas.pydata.org/docs/reference/index.html#api) is a data analysis library with a huge list of features. It is very good at holding and manipulating table data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e897ec25-52af-4e2b-9d90-e5c2e03da580",
   "metadata": {},
   "source": [
    "[jsonapi-client](https://pypi.org/project/jsonapi-client/) is a library to get formatted data from web services into python code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c3effca5-84e9-4890-baad-20baf8073f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonapi_client import Session as APISession\n",
    "from jsonapi_client import Modifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fa11014-51f3-4b5c-8e62-1256e70a2423",
   "metadata": {
    "tags": []
   },
   "source": [
    "# The MGnify API\n",
    "## Core concepts\n",
    "An [API](https://en.wikipedia.org/wiki/API \"Application programming interface\") is how your scripts (e.g. Python or R) can talk to the MGnify database.\n",
    "\n",
    "The MGnify API uses [JSON](https://en.wikipedia.org/wiki/JSON \"Javascript Object Notation\") to transfer data in a systematic way. This is human-readable and computer-readable.\n",
    "\n",
    "The particular format we use is a standard called [JSON:API](https://jsonapi.org). \n",
    "There is a Python package ([`jsonapi_client`](https://pypi.org/project/jsonapi-client/)) to make consuming this data easy. We're using it here.\n",
    "\n",
    "The MGnify API has a \"browsable interface\", which is a human-friendly way of exploring the API. The URLs for the browsable API are exactly the same as you'd use in a script or code; but when you open those URLs in a browser you see a nice interface. Find it here: [https://www.ebi.ac.uk/metagenomics/api/v1/](https://www.ebi.ac.uk/metagenomics/api/v1/).\n",
    "\n",
    "The MGnify API is \"paginated\", i.e. when you list some data you are given it in multiple pages. This is because there can sometimes by thousands of results. Thankfully `jsonapi_client` handles this for us."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa11c78-1ea9-4a98-ada5-dd82a5f50a08",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Example\n",
    "The MGnify website has a list of [\"Studies\"](https://www.ebi.ac.uk/metagenomics/browse).\n",
    "\n",
    "What the website is actually showing, is the data from an API endpoint (i.e. specific resource within the API) that lists those. It's here: [api/v1/studies](https://www.ebi.ac.uk/metagenomics/api/v1/studies). Have a look.\n",
    "\n",
    "Here is an example of some Python code, using two popular packages that let us write a short tidy piece of code:\n",
    "\n",
    "**Click into the next cell, and press `shift + return` (or click the ▶ icon on the menubar at the top) to run it.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.samples-count</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.bioproject</th>\n",
       "      <th>attributes.secondary-accession</th>\n",
       "      <th>attributes.centre-name</th>\n",
       "      <th>attributes.is-public</th>\n",
       "      <th>attributes.study-abstract</th>\n",
       "      <th>attributes.study-name</th>\n",
       "      <th>attributes.data-origination</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biomes.data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00004980</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00004980</td>\n",
       "      <td>PRJEB30236</td>\n",
       "      <td>ERP112667</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The mixed culture metagenome Third Party Annot...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-06-12T20:04:07</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae', 'type': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00004979</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00004979</td>\n",
       "      <td>PRJEB30235</td>\n",
       "      <td>ERP112666</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The mixed culture metagenome Third Party Annot...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-06-12T20:03:33</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae', 'type': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00004863</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00004863</td>\n",
       "      <td>PRJEB29830</td>\n",
       "      <td>ERP112180</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-06-12T18:37:29</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003973</td>\n",
       "      <td>338</td>\n",
       "      <td>MGYS00003973</td>\n",
       "      <td>PRJEB19645</td>\n",
       "      <td>ERP021699</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>True</td>\n",
       "      <td>Over the last decade, disease has emerged from...</td>\n",
       "      <td>Latitudinal surveys of algal-associated microo...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-04-01T12:11:56</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae', 'type': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003463</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003463</td>\n",
       "      <td>PRJEB26310</td>\n",
       "      <td>ERP108286</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-11-07T09:48:37</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003341</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003341</td>\n",
       "      <td>PRJEB26414</td>\n",
       "      <td>ERP108400</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:49:56</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003332</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003332</td>\n",
       "      <td>PRJEB26312</td>\n",
       "      <td>ERP108288</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:35:38</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003331</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003331</td>\n",
       "      <td>PRJEB26311</td>\n",
       "      <td>ERP108287</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:34:14</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003330</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003330</td>\n",
       "      <td>PRJEB26309</td>\n",
       "      <td>ERP108285</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:33:31</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003329</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003329</td>\n",
       "      <td>PRJEB26308</td>\n",
       "      <td>ERP108284</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:32:48</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003328</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003328</td>\n",
       "      <td>PRJEB26307</td>\n",
       "      <td>ERP108283</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:32:03</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003317</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYS00003317</td>\n",
       "      <td>PRJEB26275</td>\n",
       "      <td>ERP108248</td>\n",
       "      <td>EMG</td>\n",
       "      <td>True</td>\n",
       "      <td>The marine metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T18:18:32</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Red algae'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002998</td>\n",
       "      <td>22</td>\n",
       "      <td>MGYS00002998</td>\n",
       "      <td>PRJEB1154</td>\n",
       "      <td>ERP002019</td>\n",
       "      <td>CCME-COLORADO</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacteria control major nutrient cycles and dir...</td>\n",
       "      <td>The ecology of microbial communities associate...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-09-12T11:19:58</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Brown Alga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002643</td>\n",
       "      <td>96</td>\n",
       "      <td>MGYS00002643</td>\n",
       "      <td>PRJNA310780</td>\n",
       "      <td>SRP069256</td>\n",
       "      <td>University of Kiel</td>\n",
       "      <td>True</td>\n",
       "      <td>Restructuring of epibacterial communities on F...</td>\n",
       "      <td>Biological Impacts of Ocean Acidification</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2018-07-25T15:47:57</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Brown Alga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002473</td>\n",
       "      <td>69</td>\n",
       "      <td>MGYS00002473</td>\n",
       "      <td>PRJEB23005</td>\n",
       "      <td>ERP104738</td>\n",
       "      <td>Alfred Wegener Institute Helmholtz Centre for ...</td>\n",
       "      <td>True</td>\n",
       "      <td>We aimed to explore the community composition ...</td>\n",
       "      <td>Microbial diversity of the central Arctic Ocean</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-06-29T14:31:43</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae', 'type': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001287</td>\n",
       "      <td>12</td>\n",
       "      <td>MGYS00001287</td>\n",
       "      <td>PRJEB15676</td>\n",
       "      <td>ERP017509</td>\n",
       "      <td>School of Ocean Earth Sciences and Technology</td>\n",
       "      <td>True</td>\n",
       "      <td>Bacteria attached to a diatom host cell are th...</td>\n",
       "      <td>scs diatom</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2016-10-27T11:12:15</td>\n",
       "      <td>[{'id': 'root:Host-associated:Algae:Brown Alga...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       type            id  attributes.samples-count attributes.accession  \\\n",
       "0   studies  MGYS00004980                         1         MGYS00004980   \n",
       "1   studies  MGYS00004979                         1         MGYS00004979   \n",
       "2   studies  MGYS00004863                         1         MGYS00004863   \n",
       "3   studies  MGYS00003973                       338         MGYS00003973   \n",
       "4   studies  MGYS00003463                         1         MGYS00003463   \n",
       "5   studies  MGYS00003341                         1         MGYS00003341   \n",
       "6   studies  MGYS00003332                         1         MGYS00003332   \n",
       "7   studies  MGYS00003331                         1         MGYS00003331   \n",
       "8   studies  MGYS00003330                         1         MGYS00003330   \n",
       "9   studies  MGYS00003329                         1         MGYS00003329   \n",
       "10  studies  MGYS00003328                         1         MGYS00003328   \n",
       "11  studies  MGYS00003317                         1         MGYS00003317   \n",
       "12  studies  MGYS00002998                        22         MGYS00002998   \n",
       "13  studies  MGYS00002643                        96         MGYS00002643   \n",
       "14  studies  MGYS00002473                        69         MGYS00002473   \n",
       "15  studies  MGYS00001287                        12         MGYS00001287   \n",
       "\n",
       "   attributes.bioproject attributes.secondary-accession  \\\n",
       "0             PRJEB30236                      ERP112667   \n",
       "1             PRJEB30235                      ERP112666   \n",
       "2             PRJEB29830                      ERP112180   \n",
       "3             PRJEB19645                      ERP021699   \n",
       "4             PRJEB26310                      ERP108286   \n",
       "5             PRJEB26414                      ERP108400   \n",
       "6             PRJEB26312                      ERP108288   \n",
       "7             PRJEB26311                      ERP108287   \n",
       "8             PRJEB26309                      ERP108285   \n",
       "9             PRJEB26308                      ERP108284   \n",
       "10            PRJEB26307                      ERP108283   \n",
       "11            PRJEB26275                      ERP108248   \n",
       "12             PRJEB1154                      ERP002019   \n",
       "13           PRJNA310780                      SRP069256   \n",
       "14            PRJEB23005                      ERP104738   \n",
       "15            PRJEB15676                      ERP017509   \n",
       "\n",
       "                               attributes.centre-name  attributes.is-public  \\\n",
       "0                                                 EMG                  True   \n",
       "1                                                 EMG                  True   \n",
       "2                                                 EMG                  True   \n",
       "3                                              UCSDMI                  True   \n",
       "4                                                 EMG                  True   \n",
       "5                                                 EMG                  True   \n",
       "6                                                 EMG                  True   \n",
       "7                                                 EMG                  True   \n",
       "8                                                 EMG                  True   \n",
       "9                                                 EMG                  True   \n",
       "10                                                EMG                  True   \n",
       "11                                                EMG                  True   \n",
       "12                                      CCME-COLORADO                  True   \n",
       "13                                 University of Kiel                  True   \n",
       "14  Alfred Wegener Institute Helmholtz Centre for ...                  True   \n",
       "15      School of Ocean Earth Sciences and Technology                  True   \n",
       "\n",
       "                            attributes.study-abstract  \\\n",
       "0   The mixed culture metagenome Third Party Annot...   \n",
       "1   The mixed culture metagenome Third Party Annot...   \n",
       "2   The marine metagenome Third Party Annotation (...   \n",
       "3   Over the last decade, disease has emerged from...   \n",
       "4   The marine metagenome Third Party Annotation (...   \n",
       "5   The marine metagenome Third Party Annotation (...   \n",
       "6   The marine metagenome Third Party Annotation (...   \n",
       "7   The marine metagenome Third Party Annotation (...   \n",
       "8   The marine metagenome Third Party Annotation (...   \n",
       "9   The marine metagenome Third Party Annotation (...   \n",
       "10  The marine metagenome Third Party Annotation (...   \n",
       "11  The marine metagenome Third Party Annotation (...   \n",
       "12  Bacteria control major nutrient cycles and dir...   \n",
       "13  Restructuring of epibacterial communities on F...   \n",
       "14  We aimed to explore the community composition ...   \n",
       "15  Bacteria attached to a diatom host cell are th...   \n",
       "\n",
       "                                attributes.study-name  \\\n",
       "0   EMG produced TPA metagenomics assembly of the ...   \n",
       "1   EMG produced TPA metagenomics assembly of the ...   \n",
       "2   EMG produced TPA metagenomics assembly of the ...   \n",
       "3   Latitudinal surveys of algal-associated microo...   \n",
       "4   EMG produced TPA metagenomics assembly of the ...   \n",
       "5   EMG produced TPA metagenomics assembly of the ...   \n",
       "6   EMG produced TPA metagenomics assembly of the ...   \n",
       "7   EMG produced TPA metagenomics assembly of the ...   \n",
       "8   EMG produced TPA metagenomics assembly of the ...   \n",
       "9   EMG produced TPA metagenomics assembly of the ...   \n",
       "10  EMG produced TPA metagenomics assembly of the ...   \n",
       "11  EMG produced TPA metagenomics assembly of the ...   \n",
       "12  The ecology of microbial communities associate...   \n",
       "13          Biological Impacts of Ocean Acidification   \n",
       "14    Microbial diversity of the central Arctic Ocean   \n",
       "15                                         scs diatom   \n",
       "\n",
       "   attributes.data-origination attributes.last-update  \\\n",
       "0                    SUBMITTED    2019-06-12T20:04:07   \n",
       "1                    SUBMITTED    2019-06-12T20:03:33   \n",
       "2                    SUBMITTED    2019-06-12T18:37:29   \n",
       "3                    SUBMITTED    2019-04-01T12:11:56   \n",
       "4                    SUBMITTED    2018-11-07T09:48:37   \n",
       "5                    SUBMITTED    2018-10-25T18:49:56   \n",
       "6                    SUBMITTED    2018-10-25T18:35:38   \n",
       "7                    SUBMITTED    2018-10-25T18:34:14   \n",
       "8                    SUBMITTED    2018-10-25T18:33:31   \n",
       "9                    SUBMITTED    2018-10-25T18:32:48   \n",
       "10                   SUBMITTED    2018-10-25T18:32:03   \n",
       "11                   SUBMITTED    2018-10-25T18:18:32   \n",
       "12                   SUBMITTED    2018-09-12T11:19:58   \n",
       "13                   HARVESTED    2018-07-25T15:47:57   \n",
       "14                   SUBMITTED    2018-06-29T14:31:43   \n",
       "15                   SUBMITTED    2016-10-27T11:12:15   \n",
       "\n",
       "                            relationships.biomes.data  \n",
       "0   [{'id': 'root:Host-associated:Algae', 'type': ...  \n",
       "1   [{'id': 'root:Host-associated:Algae', 'type': ...  \n",
       "2   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "3   [{'id': 'root:Host-associated:Algae', 'type': ...  \n",
       "4   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "5   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "6   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "7   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "8   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "9   [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "10  [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "11  [{'id': 'root:Host-associated:Algae:Red algae'...  \n",
       "12  [{'id': 'root:Host-associated:Algae:Brown Alga...  \n",
       "13  [{'id': 'root:Host-associated:Algae:Brown Alga...  \n",
       "14  [{'id': 'root:Host-associated:Algae', 'type': ...  \n",
       "15  [{'id': 'root:Host-associated:Algae:Brown Alga...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "endpoint = \"studies\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    biome_filter = Modifier(\"lineage=root:Host-associated:Algae\")\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint, filter=biome_filter))\n",
    "    resources = pd.json_normalize(resources)\n",
    "    resources.to_csv(f\"{endpoint}.csv\")\n",
    "resources"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70c5920-17f3-4a8a-a68d-9d40d7f8ef9b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Line by line explanation\n",
    "\n",
    "```python\n",
    "### The packages were already imported, but if you wanted to use this snippet on it's own as a script you would import them like this:\n",
    "from jsonapi_client import Session as APISession\n",
    "import pandas as pd\n",
    "###\n",
    "\n",
    "\n",
    "endpoint = 'studies'\n",
    "# An \"endpoint\" is the specific resource within the API which we want to get data from. \n",
    "# It is the a URL relative to the \"server base URL\" of the API, which for MGnify is https://www.ebi.ac.uk/metagenomics/api/v1.\n",
    "# You can find the endpoints in the API Docs https://www.ebi.ac.uk/metagenomics/api/docs/ \n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    # Calling \"APISession\" is enabling a connection to the MGnify API, that can be used multiple times. \n",
    "    # The `with...as mgnify` syntax is a Python \"context\". \n",
    "    # Everything inside the `with...` block (i.e. indented below it) can use the `APISession` which we've called `mgnify` here. \n",
    "    # When the `with` block closes (the indentation stops), the connection to the API is nicely cleaned up for us.\n",
    "    \n",
    "    # Using a Modifier, we can filter the results from the API. \n",
    "    # The biome_filter will add the \"lineage=XXX\" to the query sent to the API\n",
    "    # This will be used by the API to filter the studies in the response by the biome specified in \"lineage\"\n",
    "    biome_filter = Modifier(\"lineage=root:Host-associated:Algae\")\n",
    "\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint))\n",
    "    # `map` applies a function to every element of an iterable - so do something to each thing in a list.\n",
    "    # Remember we said the API is paginated? \n",
    "    # `mgnify.iterate(endpoint)` is a very helpful function that loops over as many pages of results as there are.\n",
    "    # `lambda r: r.json` is grabbing the JSON attribute from each Super Study returned from the API.\n",
    "    # All together, this makes `resources` be a bunch of JSON representations we could loop through, each containing the data of a Super Study.\n",
    "    \n",
    "    resources = pd.json_normalize(resources)\n",
    "    # `pd` is the de-facto shorthand for the `pandas` package - you'll see it anywhere people are using pandas.\n",
    "    # The `json_normalize` function takes \"nested\" data and does its best to turn it into a table.\n",
    "    # You can throw quite strange-looking data at it and it usually does something sensible.\n",
    "    \n",
    "    resources.to_csv(f\"{endpoint}.csv\")\n",
    "    # Pandas has a built-in way of writing CSV (or TSV, etc) files, which is helpful for getting data into other tools.\n",
    "    # This writes the table-ified Super Study list to a file called `super-studies.csv`.\n",
    "    \n",
    "resources\n",
    "# In a Jupyter notebook, you can just write a variable name in a cell (or the last line of a long cell), and it will print it.\n",
    "# Jupyter knows how to display Pandas tables (actually called \"DataFrames\", because they are More Than Just Tables ™) in a pretty way.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5b2872c-9e84-4c91-8cb9-1216f2099986",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Task - Get a study from the API\n",
    "**In the cell below, complete the Python code to fetch the study _MGYS00002045_ [Study MGYS00002045 MGnify API endpoint](https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002045), and show the study data in a table.**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bece0caa-1ab1-4626-ad6f-4680d8fd334a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete this code\n",
    "resource = \n",
    "accession = \n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as    :\n",
    "    study = mgnify.get(resource, accession).resource\n",
    "    study = \n",
    "study"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27532217-e304-4d7f-b89b-7a0921833020",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Solution\n",
    "Unhide these cells to see a solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e1d16709-2677-44e8-94a4-f16e0682bb8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.bioproject</th>\n",
       "      <th>attributes.samples-count</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.secondary-accession</th>\n",
       "      <th>attributes.centre-name</th>\n",
       "      <th>attributes.is-public</th>\n",
       "      <th>attributes.study-abstract</th>\n",
       "      <th>attributes.study-name</th>\n",
       "      <th>attributes.data-origination</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biomes.data</th>\n",
       "      <th>relationships.studies.data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002045</td>\n",
       "      <td>PRJEB22528</td>\n",
       "      <td>325</td>\n",
       "      <td>MGYS00002045</td>\n",
       "      <td>ERP104209</td>\n",
       "      <td>EMBL-EBI</td>\n",
       "      <td>True</td>\n",
       "      <td>The human gut metagenome Third Party Annotatio...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2021-10-22T18:11:17</td>\n",
       "      <td>[{'id': 'root:Host-associated:Human:Digestive ...</td>\n",
       "      <td>[{'id': 'MGYS00000633', 'type': 'studies'}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type            id attributes.bioproject  attributes.samples-count  \\\n",
       "0  studies  MGYS00002045            PRJEB22528                       325   \n",
       "\n",
       "  attributes.accession attributes.secondary-accession attributes.centre-name  \\\n",
       "0         MGYS00002045                      ERP104209               EMBL-EBI   \n",
       "\n",
       "   attributes.is-public                          attributes.study-abstract  \\\n",
       "0                  True  The human gut metagenome Third Party Annotatio...   \n",
       "\n",
       "                               attributes.study-name  \\\n",
       "0  EMG produced TPA metagenomics assembly of the ...   \n",
       "\n",
       "  attributes.data-origination attributes.last-update  \\\n",
       "0                   SUBMITTED    2021-10-22T18:11:17   \n",
       "\n",
       "                           relationships.biomes.data  \\\n",
       "0  [{'id': 'root:Host-associated:Human:Digestive ...   \n",
       "\n",
       "                    relationships.studies.data  \n",
       "0  [{'id': 'MGYS00000633', 'type': 'studies'}]  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resource = \"studies\"\n",
    "accession = \"MGYS00002045\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    study = mgnify.get(resource, accession).resource\n",
    "    study = pd.json_normalize(study.json)\n",
    "study"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c2e909ea68d00cf590af4aa2de30821786e45b8e11ea607135c2643544907b22"
  },
  "kernelspec": {
   "display_name": "Python 3.10.0 64-bit ('mgnify-ebi-2021': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
