{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3be1d41e-2365-4c09-84c2-84e177cf254c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Metagenomics Bioinformatics Course - EBI MGnify 2021\n",
    "## MGnify Genomes resource - Metagenomic Assembled Genomes Catalogues - Practical exercise\n",
    "\n",
    "### Aims\n",
    "In this exercise, we will learn how to use the [Genomes resource within MGnify](https://www.ebi.ac.uk/metagenomics/browse#genomes).\n",
    "\n",
    "- Discover the available data on the MGnify website\n",
    "- Use two search mechanisms (search by _gene_ and search by _genome_)\n",
    "- Learn how to use the MGnify API to fetch data using scripts or analysis notebooks\n",
    "- Use the _genome_ search mechanism via the API, to compare your own MAGs against a MGnify catalogue and see whether they are novel\n",
    "\n",
    "### How this works\n",
    "This file is a [Jupyter Notebook](https://jupyter.org). \n",
    "It has instructions, and also code cells. The code cells are connected to Python, and you can run all of the code in a cell by pressing Play (▶) icon in the top bar, or pressing `shift + return`.\n",
    "The code libraries you should need are already installed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9704e24e-b25c-4c5e-b02d-ecd4f23d085c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a639807-3595-43a8-8d36-eaddd42adb23",
   "metadata": {
    "tags": []
   },
   "source": [
    "[pandas](https://pandas.pydata.org/docs/reference/index.html#api) is a data analysis library with a huge list of features. It is very good at holding and manipulating table data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "51cb916f-02e9-46db-9013-01d63516ee96",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e897ec25-52af-4e2b-9d90-e5c2e03da580",
   "metadata": {},
   "source": [
    "[jsonapi-client](https://pypi.org/project/jsonapi-client/) is a library to get formatted data from web services into python code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c3effca5-84e9-4890-baad-20baf8073f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonapi_client import Session as APISession"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fa11014-51f3-4b5c-8e62-1256e70a2423",
   "metadata": {
    "tags": []
   },
   "source": [
    "# The MGnify API\n",
    "## Core concepts\n",
    "An [API](https://en.wikipedia.org/wiki/API \"Application programming interface\") is how your scripts (e.g. Python or R) can talk to the MGnify database.\n",
    "\n",
    "The MGnify API uses [JSON](https://en.wikipedia.org/wiki/JSON \"Javascript Object Notation\") to transfer data in a systematic way. This is human-readable and computer-readable.\n",
    "\n",
    "The particular format we use is a standard called [JSON:API](https://jsonapi.org). \n",
    "There is a Python package ([`jsonapi_client`](https://pypi.org/project/jsonapi-client/)) to make consuming this data easy. We're using it here.\n",
    "\n",
    "The MGnify API has a \"browsable interface\", which is a human-friendly way of exploring the API. The URLs for the browsable API are exactly the same as you'd use in a script or code; but when you open those URLs in a browser you see a nice interface. Find it here: [https://www.ebi.ac.uk/metagenomics/api/v1/](https://www.ebi.ac.uk/metagenomics/api/v1/).\n",
    "\n",
    "The MGnify API is \"paginated\", i.e. when you list some data you are given it in multiple pages. This is because there can sometimes by thousands of results. Thankfully `jsonapi_client` handles this for us."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa11c78-1ea9-4a98-ada5-dd82a5f50a08",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Example\n",
    "The MGnify website has a list of [\"Super Studies\"](https://www.ebi.ac.uk/metagenomics/browse) (collections of studies that together represent major research efforts or collaborations).\n",
    "\n",
    "What the website is actually showing, is the data from an API endpoint (i.e. specific resource within the API) that lists those. It's here: [api/v1/super-studies](https://www.ebi.ac.uk/metagenomics/api/v1/super-studies). Have a look.\n",
    "\n",
    "Here is an example of some Python code, using two popular packages that let us write a short tidy piece of code:\n",
    "\n",
    "**Click into the next cell, and press `shift + return` (or click the ▶ icon on the menubar at the top) to run it.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8b684884-80b1-4bef-a5dc-634f6e7bfdba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.super-study-id</th>\n",
       "      <th>attributes.title</th>\n",
       "      <th>attributes.url-slug</th>\n",
       "      <th>attributes.description</th>\n",
       "      <th>attributes.image-url</th>\n",
       "      <th>attributes.biomes-count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>super-studies</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Tara Oceans</td>\n",
       "      <td>tara-oceans</td>\n",
       "      <td>The Tara Oceans expedition (Karsenti et al. 20...</td>\n",
       "      <td>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>super-studies</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Earth Microbiome Project</td>\n",
       "      <td>earth-microbiome-project</td>\n",
       "      <td>The Earth Microbiome Project is now available ...</td>\n",
       "      <td>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>super-studies</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NASA GeneLab Microbiome (MANGO)</td>\n",
       "      <td>nasa-genelab-microbiome-mango</td>\n",
       "      <td>Project MANGO provides access to the microbiom...</td>\n",
       "      <td>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>super-studies</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>HoloFood</td>\n",
       "      <td>holofood</td>\n",
       "      <td>Holistic approach to improve the efficiency of...</td>\n",
       "      <td>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>super-studies</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>Malaspina</td>\n",
       "      <td>malaspina</td>\n",
       "      <td>The Malaspina circumnavigation expedition was ...</td>\n",
       "      <td>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>super-studies</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>AtlantECO</td>\n",
       "      <td>atlanteco</td>\n",
       "      <td>The EU-funded AtlantECO project aims to develo...</td>\n",
       "      <td>data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            type id  attributes.super-study-id  \\\n",
       "0  super-studies  1                          1   \n",
       "1  super-studies  2                          2   \n",
       "2  super-studies  3                          3   \n",
       "3  super-studies  4                          4   \n",
       "4  super-studies  5                          5   \n",
       "5  super-studies  6                          6   \n",
       "\n",
       "                  attributes.title            attributes.url-slug  \\\n",
       "0                      Tara Oceans                    tara-oceans   \n",
       "1         Earth Microbiome Project       earth-microbiome-project   \n",
       "2  NASA GeneLab Microbiome (MANGO)  nasa-genelab-microbiome-mango   \n",
       "3                         HoloFood                       holofood   \n",
       "4                        Malaspina                      malaspina   \n",
       "5                        AtlantECO                      atlanteco   \n",
       "\n",
       "                              attributes.description  \\\n",
       "0  The Tara Oceans expedition (Karsenti et al. 20...   \n",
       "1  The Earth Microbiome Project is now available ...   \n",
       "2  Project MANGO provides access to the microbiom...   \n",
       "3  Holistic approach to improve the efficiency of...   \n",
       "4  The Malaspina circumnavigation expedition was ...   \n",
       "5  The EU-funded AtlantECO project aims to develo...   \n",
       "\n",
       "                                attributes.image-url  attributes.biomes-count  \n",
       "0  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...                        0  \n",
       "1  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...                        0  \n",
       "2  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...                        0  \n",
       "3  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...                        0  \n",
       "4  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...                        0  \n",
       "5  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...                        0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "endpoint = 'super-studies'\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint))\n",
    "    resources = pd.json_normalize(resources)\n",
    "    resources.to_csv(f\"{endpoint}.csv\")\n",
    "resources"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70c5920-17f3-4a8a-a68d-9d40d7f8ef9b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Line by line explanation\n",
    "\n",
    "```python\n",
    "### The packages were already imported, but if you wanted to use this snippet on it's own as a script you would import them like this:\n",
    "from jsonapi_client import Session as APISession\n",
    "import pandas as pd\n",
    "###\n",
    "\n",
    "\n",
    "endpoint = 'super-studies'\n",
    "# An \"endpoint\" is the specific resource within the API which we want to get data from. \n",
    "# It is the a URL relative to the \"server base URL\" of the API, which for MGnify is https://www.ebi.ac.uk/metagenomics/api/v1.\n",
    "# You can find the endpoints in the API Docs https://www.ebi.ac.uk/metagenomics/api/docs/ \n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    # Calling \"APISession\" is enabling a connection to the MGnify API, that can be used multiple times. \n",
    "    # The `with...as mgnify` syntax is a Python \"context\". \n",
    "    # Everything inside the `with...` block (i.e. indented below it) can use the `APISession` which we've called `mgnify` here. \n",
    "    # When the `with` block closes (the indentation stops), the connection to the API is nicely cleaned up for us.\n",
    "    \n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint))\n",
    "    # `map` applies a function to every element of an iterable - so do something to each thing in a list.\n",
    "    # Remember we said the API is paginated? \n",
    "    # `mgnify.iterate(endpoint)` is a very helpful function that loops over as many pages of results as there are.\n",
    "    # `lambda r: r.json` is grabbing the JSON attribute from each Super Study returned from the API.\n",
    "    # All together, this makes `resources` be a bunch of JSON representations we could loop through, each containing the data of a Super Study.\n",
    "    \n",
    "    resources = pd.json_normalize(resources)\n",
    "    # `pd` is the de-facto shorthand for the `pandas` package - you'll see it anywhere people are using pandas.\n",
    "    # The `json_normalize` function takes \"nested\" data and does its best to turn it into a table.\n",
    "    # You can throw quite strange-looking data at it and it usually does something sensible.\n",
    "    \n",
    "    resources.to_csv(f\"{endpoint}.csv\")\n",
    "    # Pandas has a built-in way of writing CSV (or TSV, etc) files, which is helpful for getting data into other tools.\n",
    "    # This writes the table-ified Super Study list to a file called `super-studies.csv`.\n",
    "    \n",
    "resources\n",
    "# In a Jupyter notebook, you can just write a variable name in a cell (or the last line of a long cell), and it will print it.\n",
    "# Jupyter knows how to display Pandas tables (actually called \"DataFrames\", because they are More Than Just Tables ™) in a pretty way.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5b2872c-9e84-4c91-8cb9-1216f2099986",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Task 1 - list Genome Catalogues\n",
    "**In the cell below, complete the Python code to fetch the list of [Genome Catalogues from the MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1/genome-catalogues), and show them in a table.**\n",
    "\n",
    "(Note that there may only be one catalogue in the list right now, that is correct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bece0caa-1ab1-4626-ad6f-4680d8fd334a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete this code\n",
    "\n",
    "endpoint = \n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    catalogues = \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "catalogues"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27532217-e304-4d7f-b89b-7a0921833020",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Solution\n",
    "Unhide these cells to see a solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e1d16709-2677-44e8-94a4-f16e0682bb8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.name</th>\n",
       "      <th>attributes.description</th>\n",
       "      <th>attributes.protein-catalogue-name</th>\n",
       "      <th>attributes.protein-catalogue-description</th>\n",
       "      <th>attributes.genome-count</th>\n",
       "      <th>attributes.version</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biome.data.id</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>Unified Human Gastrointestinal Genome (UHGG) 1.0</td>\n",
       "      <td>4,644 species-level prokaryotic genomes corres...</td>\n",
       "      <td>The Unified Human Gastrointestinal Protein cat...</td>\n",
       "      <td>The Unified Human Gastrointestinal Protein (UH...</td>\n",
       "      <td>4644</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2021-09-30T11:27:14.126429</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                type              id  \\\n",
       "0  genome-catalogues  human-gut-v1-0   \n",
       "\n",
       "                                    attributes.name  \\\n",
       "0  Unified Human Gastrointestinal Genome (UHGG) 1.0   \n",
       "\n",
       "                              attributes.description  \\\n",
       "0  4,644 species-level prokaryotic genomes corres...   \n",
       "\n",
       "                   attributes.protein-catalogue-name  \\\n",
       "0  The Unified Human Gastrointestinal Protein cat...   \n",
       "\n",
       "            attributes.protein-catalogue-description  attributes.genome-count  \\\n",
       "0  The Unified Human Gastrointestinal Protein (UH...                     4644   \n",
       "\n",
       "  attributes.version      attributes.last-update  \\\n",
       "0                1.0  2021-09-30T11:27:14.126429   \n",
       "\n",
       "                         relationships.biome.data.id  \\\n",
       "0  root:Host-associated:Human:Digestive system:La...   \n",
       "\n",
       "  relationships.biome.data.type  \n",
       "0                        biomes  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "endpoint = 'genome-catalogues'\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    catalogues = map(lambda r: r.json, mgnify.iterate(endpoint))\n",
    "    catalogues = pd.json_normalize(catalogues)\n",
    "catalogues"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa460ee7-f93d-4388-b044-6c20e07501e2",
   "metadata": {},
   "source": [
    "# Task 2 - list Genomes\n",
    "Each catalogue contains a much larger list of Genomes.\n",
    "**In the cell below, complete the Python code to fetch the list of [Genomes from the MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1/genome-catalogues), and show them in a table.**\n",
    "\n",
    "(Note that there are quite a lot of pages of data, so this will take a minute to run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88521094-fc50-4e4f-b0fc-8651a1207c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "catalogue_id = catalogues.id[0]\n",
    "endpoint = f'genome-catalogues/{catalogue_id}/genomes'  # a Python f-string inserts the value of a variable into the string where that variable name appears inside {..}\n",
    "\n",
    "with           as mgnify:\n",
    "    genomes = \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7e15f6c-5ab8-4d0c-a921-405cc22b0a0e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Solution\n",
    "Unhide these cells to see a solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e91c0456-3c0d-4e54-bddb-fac3d233e56d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.genome-id</th>\n",
       "      <th>attributes.geographic-range</th>\n",
       "      <th>attributes.geographic-origin</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.ena-sample-accession</th>\n",
       "      <th>attributes.ena-study-accession</th>\n",
       "      <th>attributes.length</th>\n",
       "      <th>attributes.num-contigs</th>\n",
       "      <th>...</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "      <th>relationships.catalogue.data.id</th>\n",
       "      <th>relationships.catalogue.data.type</th>\n",
       "      <th>attributes.ncbi-genome-accession</th>\n",
       "      <th>attributes.ncbi-sample-accession</th>\n",
       "      <th>attributes.ncbi-study-accession</th>\n",
       "      <th>attributes.cmseq</th>\n",
       "      <th>attributes.ena-genome-accession</th>\n",
       "      <th>attributes.img-genome-accession</th>\n",
       "      <th>attributes.patric-genome-accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-00001</td>\n",
       "      <td>648</td>\n",
       "      <td>[North America, Europe]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-00001</td>\n",
       "      <td>ERS370061</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>3219614</td>\n",
       "      <td>137</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-00002</td>\n",
       "      <td>383</td>\n",
       "      <td>[North America, Europe, Asia, Oceania, South A...</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-00002</td>\n",
       "      <td>ERS370064</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>4433090</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-00003</td>\n",
       "      <td>2319</td>\n",
       "      <td>[North America, Europe, Asia, Oceania, South A...</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-00003</td>\n",
       "      <td>ERS370070</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>3229507</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-00004</td>\n",
       "      <td>4150</td>\n",
       "      <td>[North America, Europe, Asia]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-00004</td>\n",
       "      <td>ERS370072</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>3698872</td>\n",
       "      <td>105</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-00005</td>\n",
       "      <td>2855</td>\n",
       "      <td>[North America, Europe]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-00005</td>\n",
       "      <td>ERS417217</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>3930422</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4639</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-04640</td>\n",
       "      <td>498</td>\n",
       "      <td>[Europe, Asia]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-04640</td>\n",
       "      <td>ERS436824</td>\n",
       "      <td>ERP005534</td>\n",
       "      <td>2596429</td>\n",
       "      <td>276</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4640</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-04641</td>\n",
       "      <td>995</td>\n",
       "      <td>[]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-04641</td>\n",
       "      <td>ERS436802</td>\n",
       "      <td>ERP005534</td>\n",
       "      <td>1772088</td>\n",
       "      <td>174</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4641</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-04642</td>\n",
       "      <td>464</td>\n",
       "      <td>[North America, Europe, Asia]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-04642</td>\n",
       "      <td>ERS436800</td>\n",
       "      <td>ERP005534</td>\n",
       "      <td>2001820</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4642</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-04643</td>\n",
       "      <td>1136</td>\n",
       "      <td>[Africa, Asia, Europe, North America, Oceania,...</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-04643</td>\n",
       "      <td>ERS436826</td>\n",
       "      <td>ERP005534</td>\n",
       "      <td>1975847</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4643</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG-HGUT-04644</td>\n",
       "      <td>3094</td>\n",
       "      <td>[North America, Europe, Asia]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG-HGUT-04644</td>\n",
       "      <td>ERS436840</td>\n",
       "      <td>ERP005534</td>\n",
       "      <td>2240162</td>\n",
       "      <td>237</td>\n",
       "      <td>...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>human-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4644 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         type               id  attributes.genome-id  \\\n",
       "0     genomes  MGYG-HGUT-00001                   648   \n",
       "1     genomes  MGYG-HGUT-00002                   383   \n",
       "2     genomes  MGYG-HGUT-00003                  2319   \n",
       "3     genomes  MGYG-HGUT-00004                  4150   \n",
       "4     genomes  MGYG-HGUT-00005                  2855   \n",
       "...       ...              ...                   ...   \n",
       "4639  genomes  MGYG-HGUT-04640                   498   \n",
       "4640  genomes  MGYG-HGUT-04641                   995   \n",
       "4641  genomes  MGYG-HGUT-04642                   464   \n",
       "4642  genomes  MGYG-HGUT-04643                  1136   \n",
       "4643  genomes  MGYG-HGUT-04644                  3094   \n",
       "\n",
       "                            attributes.geographic-range  \\\n",
       "0                               [North America, Europe]   \n",
       "1     [North America, Europe, Asia, Oceania, South A...   \n",
       "2     [North America, Europe, Asia, Oceania, South A...   \n",
       "3                         [North America, Europe, Asia]   \n",
       "4                               [North America, Europe]   \n",
       "...                                                 ...   \n",
       "4639                                     [Europe, Asia]   \n",
       "4640                                                 []   \n",
       "4641                      [North America, Europe, Asia]   \n",
       "4642  [Africa, Asia, Europe, North America, Oceania,...   \n",
       "4643                      [North America, Europe, Asia]   \n",
       "\n",
       "     attributes.geographic-origin attributes.accession  \\\n",
       "0                          Europe      MGYG-HGUT-00001   \n",
       "1                          Europe      MGYG-HGUT-00002   \n",
       "2                          Europe      MGYG-HGUT-00003   \n",
       "3                          Europe      MGYG-HGUT-00004   \n",
       "4                          Europe      MGYG-HGUT-00005   \n",
       "...                           ...                  ...   \n",
       "4639                       Europe      MGYG-HGUT-04640   \n",
       "4640                       Europe      MGYG-HGUT-04641   \n",
       "4641                       Europe      MGYG-HGUT-04642   \n",
       "4642                       Europe      MGYG-HGUT-04643   \n",
       "4643                       Europe      MGYG-HGUT-04644   \n",
       "\n",
       "     attributes.ena-sample-accession attributes.ena-study-accession  \\\n",
       "0                          ERS370061                      ERP105624   \n",
       "1                          ERS370064                      ERP105624   \n",
       "2                          ERS370070                      ERP105624   \n",
       "3                          ERS370072                      ERP105624   \n",
       "4                          ERS417217                      ERP105624   \n",
       "...                              ...                            ...   \n",
       "4639                       ERS436824                      ERP005534   \n",
       "4640                       ERS436802                      ERP005534   \n",
       "4641                       ERS436800                      ERP005534   \n",
       "4642                       ERS436826                      ERP005534   \n",
       "4643                       ERS436840                      ERP005534   \n",
       "\n",
       "      attributes.length  attributes.num-contigs  ...  \\\n",
       "0               3219614                     137  ...   \n",
       "1               4433090                     100  ...   \n",
       "2               3229507                      35  ...   \n",
       "3               3698872                     105  ...   \n",
       "4               3930422                      32  ...   \n",
       "...                 ...                     ...  ...   \n",
       "4639            2596429                     276  ...   \n",
       "4640            1772088                     174  ...   \n",
       "4641            2001820                      23  ...   \n",
       "4642            1975847                      11  ...   \n",
       "4643            2240162                     237  ...   \n",
       "\n",
       "      relationships.biome.data.type  relationships.catalogue.data.id  \\\n",
       "0                            biomes                   human-gut-v1-0   \n",
       "1                            biomes                   human-gut-v1-0   \n",
       "2                            biomes                   human-gut-v1-0   \n",
       "3                            biomes                   human-gut-v1-0   \n",
       "4                            biomes                   human-gut-v1-0   \n",
       "...                             ...                              ...   \n",
       "4639                         biomes                   human-gut-v1-0   \n",
       "4640                         biomes                   human-gut-v1-0   \n",
       "4641                         biomes                   human-gut-v1-0   \n",
       "4642                         biomes                   human-gut-v1-0   \n",
       "4643                         biomes                   human-gut-v1-0   \n",
       "\n",
       "     relationships.catalogue.data.type  attributes.ncbi-genome-accession  \\\n",
       "0                    genome-catalogues                               NaN   \n",
       "1                    genome-catalogues                               NaN   \n",
       "2                    genome-catalogues                               NaN   \n",
       "3                    genome-catalogues                               NaN   \n",
       "4                    genome-catalogues                               NaN   \n",
       "...                                ...                               ...   \n",
       "4639                 genome-catalogues                               NaN   \n",
       "4640                 genome-catalogues                               NaN   \n",
       "4641                 genome-catalogues                               NaN   \n",
       "4642                 genome-catalogues                               NaN   \n",
       "4643                 genome-catalogues                               NaN   \n",
       "\n",
       "      attributes.ncbi-sample-accession  attributes.ncbi-study-accession  \\\n",
       "0                                  NaN                              NaN   \n",
       "1                                  NaN                              NaN   \n",
       "2                                  NaN                              NaN   \n",
       "3                                  NaN                              NaN   \n",
       "4                                  NaN                              NaN   \n",
       "...                                ...                              ...   \n",
       "4639                               NaN                              NaN   \n",
       "4640                               NaN                              NaN   \n",
       "4641                               NaN                              NaN   \n",
       "4642                               NaN                              NaN   \n",
       "4643                               NaN                              NaN   \n",
       "\n",
       "      attributes.cmseq  attributes.ena-genome-accession  \\\n",
       "0                  NaN                              NaN   \n",
       "1                  NaN                              NaN   \n",
       "2                  NaN                              NaN   \n",
       "3                  NaN                              NaN   \n",
       "4                  NaN                              NaN   \n",
       "...                ...                              ...   \n",
       "4639              1.09                              NaN   \n",
       "4640              0.20                              NaN   \n",
       "4641              0.05                              NaN   \n",
       "4642              0.05                              NaN   \n",
       "4643              0.60                              NaN   \n",
       "\n",
       "      attributes.img-genome-accession  attributes.patric-genome-accession  \n",
       "0                                 NaN                                 NaN  \n",
       "1                                 NaN                                 NaN  \n",
       "2                                 NaN                                 NaN  \n",
       "3                                 NaN                                 NaN  \n",
       "4                                 NaN                                 NaN  \n",
       "...                               ...                                 ...  \n",
       "4639                              NaN                                 NaN  \n",
       "4640                              NaN                                 NaN  \n",
       "4641                              NaN                                 NaN  \n",
       "4642                              NaN                                 NaN  \n",
       "4643                              NaN                                 NaN  \n",
       "\n",
       "[4644 rows x 45 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "catalogue_id = catalogues.id[0]\n",
    "endpoint = f'genome-catalogues/{catalogue_id}/genomes'  # a Python f-string inserts the value of a variable into the string where that variable name appears inside {..}\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    genomes = map(lambda r: r.json, mgnify.iterate(endpoint))\n",
    "    genomes = pd.json_normalize(genomes)\n",
    "genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "484226fd-59f5-4354-9486-0c97e085848a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mgnify-ebi-2021",
   "language": "python",
   "name": "mgnify-ebi-2021"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
